### PR TITLE
feat(generic): add stacktrace API

### PR DESCRIPTION
**Pull request by Wescoeur**

### CODE REVIEW

File: src/stacktrace.c [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+#define PACKAGE 1
+#include <bfd.h>
+
+#include "xcp-ng/generic/io.h"
+#include "xcp-ng/generic/stacktrace.h"
+
+// =============================================================================
+
+// TODO: Remove me.
```
Comment (benjamreis) [inline src/stacktrace.c]: Do it?

File: src/stacktrace.c [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+typedef struct SymbolInfo {
+  const char *filename;
+  const char *function;
+  uint line;
+  const void *physAddress;
+} SymbolInfo;
+
+static int symbol_info_to_str (const SymbolInfo *info, char *buf, size_t size) {
+  const char *function = info->function;
+  if (!function || *function == '\0')
```
Comment (benjamreis) [inline src/stacktrace.c:53]: Maybe use a ternary ?
```c
const char *function = (info->function && *info->function != '\0') ? info->function : "??";
```

File: src/stacktrace.c [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    function = "??";
+
+  const char *filename = info->filename;
+  if (!filename || *filename == '\0')
```
Comment (benjamreis) [inline src/stacktrace.c:57]: Same as above.
```c
const char *filename = (info->filename && *info->filename != '\0') ? info->filename : "??";
```

**Comment by benjamreis** [comment]
Add header in `generic.h`.
Maybe add comments in `stacktrace.h` to explain how to use it.

**Comment by benjamreis** [comment]
Maybe add comments in `stacktrace.h` to explain how to use it.

File: src/stacktrace.c (hunk 1 above)
Comment (Wescoeur) [reply src/stacktrace.c]: WIP PR. :wink:

File: src/stacktrace.c [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+static char *physical_addr_to_str_from_file (const char *filename, const void *physAddress) {
+  bfd *bin = bfd_openr(filename, NULL);
+  if (!bin)
+    return NULL;
+
+  asymbol **symbols;
+
+  if (
+    bfd_check_format(bin, bfd_archive) ||
```
Comment (benjamreis) [inline src/stacktrace.c]: Remove this line.

File: src/stacktrace.c (hunk 1 above)
Comment (benjamreis) [reply src/stacktrace.c]: :+1:

File: src/stacktrace.c (hunk 2 above)
Comment (Wescoeur) [reply src/stacktrace.c:53]: I'm not sure it's more readable like this.
I store `info->function` in `function` to avoid repetitions.

**Comment by Wescoeur** [comment]
> Add header in `generic.h`.
> Maybe add comments in `stacktrace.h` to explain how to use it.

It's similar to the GNU `backtrace` API:
```C
#include <execinfo.h>

int backtrace(void **buffer, int size);
char **backtrace_symbols(void *const *buffer, int size);
void backtrace_symbols_fd(void *const *buffer, int size, int fd);
```

I don't know what to say in a documentation but I can try to explain the basis. :thinking:

**Comment by benjamreis** [comment]
Yep, just a few line might be enough :)

**Pull request merged**